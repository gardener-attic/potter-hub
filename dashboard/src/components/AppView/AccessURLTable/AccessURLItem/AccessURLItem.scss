.ServiceItem {
  display: inline-block;
  color: #fff;
  border-radius: 16px;
  background-color: #5f6369;
  &--with-link {
    background-color: #1598cb;
    a {
      color: inherit;
    }
    &:hover {
      background-color: #f58220;
      a {
        color: inherit;
      }
    }
  }
}
