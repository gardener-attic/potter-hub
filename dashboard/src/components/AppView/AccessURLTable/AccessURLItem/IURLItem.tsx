export interface IURLItem {
  name: string;
  type: string;
  isLink: boolean;
  URLs: string[];
}
