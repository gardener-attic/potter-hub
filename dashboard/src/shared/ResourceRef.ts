import { Kube } from "./Kube";
import { ResourceKind } from "./ResourceKinds";
import { IResource } from "./types";

// ResourceRef defines a reference to a namespaced Kubernetes API Object and
// provides helpers to retrieve the resource URL
class ResourceRef {
  public apiVersion: string;
  public kind: ResourceKind;
  public name: string;
  public namespace: string;
  public isOnResourceCluster: boolean;

  // Creates a new ResourceRef instance from an existing IResource. Provide
  // defaultNamespace to set if the IResource doesn't specify a namespace.
  // TODO: add support for cluster-scoped resources, or add a ClusterResourceRef
  // class.
  constructor(r: IResource, defaultNamespace?: string, isOnResourceCluster?: boolean) {
    this.apiVersion = r.apiVersion;
    this.kind = r.kind;
    this.name = r.metadata.name;
    const namespace = r.metadata.namespace || defaultNamespace;
    if (!namespace) {
      throw new Error(`Namespace missing for resource ${this.name}, define a default namespace`);
    }
    this.namespace = namespace;
    isOnResourceCluster
      ? (this.isOnResourceCluster = isOnResourceCluster)
      : (this.isOnResourceCluster = false);
    return this;
  }

  // Gets a full resource URL for the referenced resource
  public getResourceURL() {
    return Kube.getResourceURL(
      this.apiVersion,
      Kube.resourcePlural(this.kind),
      this.isOnResourceCluster,
      this.namespace,
      this.name,
    );
  }

  public watchResourceURL() {
    return Kube.watchResourceURL(
      this.apiVersion,
      Kube.resourcePlural(this.kind),
      this.isOnResourceCluster,
      this.namespace,
      this.name,
    );
  }

  public getResource() {
    return Kube.getResource(
      this.apiVersion,
      Kube.resourcePlural(this.kind),
      this.isOnResourceCluster,
      this.namespace,
      this.name,
    );
  }

  // Opens and returns a WebSocket for the requested resource. Note: it is
  // important that this socket be properly closed when no longer needed. The
  // returned WebSocket can be attached to an event listener to read data from
  // the socket.
  public watchResource() {
    return Kube.watchResource(
      this.apiVersion,
      Kube.resourcePlural(this.kind),
      this.isOnResourceCluster,
      this.namespace,
      this.name,
    );
  }
}

export default ResourceRef;
