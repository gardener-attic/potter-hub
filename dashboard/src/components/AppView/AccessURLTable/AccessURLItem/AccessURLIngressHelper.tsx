import { IIngressSpec, IIngressTLS, IResource } from "shared/types";
import { IURLItem } from "./IURLItem";

// URLs returns the list of URLs obtained from the service status
function URLs(ingress: IResource): string[] {
  const spec = ingress.spec as IIngressSpec;
  const res: string[] = [];
  if (spec.rules) {
    spec.rules.forEach(r => {
      if (r.http && r.http.paths.length > 0) {
        r.http.paths.forEach(p => {
          res.push(getURL(r.host, spec.tls, p.path));
        });
      } else {
        res.push(getURL(r.host, spec.tls));
      }
    });
  }
  return res;
}

// getURL returns a full URL based on a hostname, a TLS configuration and a optional path
function getURL(hostname: string, tls?: IIngressTLS[], path?: string) {
  // If the hostname is configured within the TLS hosts it will use HTTPS
  const protocol =
    tls &&
    tls.some(tlsRule => {
      return tlsRule.hosts && tlsRule.hosts.indexOf(hostname) > -1;
    })
      ? "https"
      : "http";
  return `${protocol}://${hostname}${path || ""}`;
}

export function GetURLItemFromIngress(ingress: IResource) {
  return {
    name: ingress.metadata.name,
    type: "Ingress",
    isLink: true,
    URLs: URLs(ingress),
  } as IURLItem;
}
