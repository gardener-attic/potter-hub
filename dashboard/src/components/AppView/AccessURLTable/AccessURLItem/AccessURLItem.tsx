import * as React from "react";

import "./AccessURLItem.css";
import { IURLItem } from "./IURLItem";

interface IAccessURLItem {
  URLItem: IURLItem;
}

const AccessURLItem: React.SFC<IAccessURLItem> = props => {
  const { URLItem } = props;
  return (
    <tr>
      <td>{URLItem.name}</td>
      <td>{URLItem.type}</td>
      <td>
        {URLItem.URLs.map(l => (
          <span
            key={l}
            className={`ServiceItem ${
              URLItem.isLink ? "ServiceItem--with-link" : ""
            } type-small margin-r-small padding-tiny padding-h-normal`}
          >
            {URLItem.isLink ? (
              <a className="padding-tiny padding-h-normal" href={l} target="_blank" rel="noopener noreferrer">
                {l}
              </a>
            ) : (
              l
            )}
          </span>
        ))}
      </td>
    </tr>
  );
};

export default AccessURLItem;
