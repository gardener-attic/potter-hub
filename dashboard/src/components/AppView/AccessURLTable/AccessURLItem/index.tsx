import AccessURLItem from "./AccessURLItem";

export default AccessURLItem;
