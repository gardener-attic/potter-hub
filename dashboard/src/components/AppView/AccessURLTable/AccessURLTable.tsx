import * as React from "react";

import ResourceRef from "shared/ResourceRef";
import LoadingWrapper, { LoaderType } from "../../../components/LoadingWrapper";
import { IK8sList, IKubeItem, IResource, IServiceSpec } from "../../../shared/types";
import isSomeResourceLoading from "../helpers";
import AccessURLItem from "./AccessURLItem";
import { GetURLItemFromIngress } from "./AccessURLItem/AccessURLIngressHelper";
import { GetURLItemFromService } from "./AccessURLItem/AccessURLServiceHelper";

interface IAccessURLTableProps {
  services: Array<IKubeItem<IResource>>;
  ingresses: Array<IKubeItem<IResource>>;
  ingressRefs: ResourceRef[];
  getResource: (r: ResourceRef) => void;
}

class AccessURLTable extends React.Component<IAccessURLTableProps> {
  public componentDidMount() {
    this.fetchIngresses();
  }

  public componentDidUpdate(prevProps: IAccessURLTableProps) {
    if (prevProps.ingressRefs.length !== this.props.ingressRefs.length) {
      this.fetchIngresses();
    }
  }

  public render() {
    const { ingresses, services } = this.props;
    if (isSomeResourceLoading(ingresses.concat(services))) {
      return <LoadingWrapper type={LoaderType.Placeholder} />;
    }
    if (!this.hasItems(services, ingresses)) {
      return null;
    }
    return (
      <React.Fragment>
        <h6>Access URLs</h6>
        {this.accessTableSection()}
      </React.Fragment>
    );
  }

  private publicServices(): Array<IKubeItem<IResource>> {
    const { services } = this.props;
    const publicServices: Array<IKubeItem<IResource>> = [];
    services.forEach(s => {
      if (s.item) {
        const spec = s.item.spec as IServiceSpec;
        if (spec.type === "LoadBalancer") {
          publicServices.push(s);
        }
      }
    });
    return publicServices;
  }

  private accessTableSection() {
    const { ingresses } = this.props;
    let accessTableSection = <p>The current application does not expose a public URL.</p>;
    const publicServices = this.publicServices();
    if (publicServices.length > 0 || ingresses.length > 0) {
      accessTableSection = (
        <React.Fragment>
          <table>
            <thead>
              <tr>
                <th>RESOURCE</th>
                <th>TYPE</th>
                <th>URL</th>
              </tr>
            </thead>
            <tbody>
              {ingresses.map(i => this.renderTableEntry(i))}
              {publicServices.map(s => this.renderTableEntry(s))}
            </tbody>
          </table>
        </React.Fragment>
      );
    }
    return accessTableSection;
  }

  private renderTableEntry(i: IKubeItem<IResource>) {
    if (i.error) {
      return (
        <tr key={i.error.message}>
          <td colSpan={3}>Error: {i.error.message}</td>
        </tr>
      );
    }
    if (i.item) {
      const urlItem =
        i.item.kind === "Ingress" ? GetURLItemFromIngress(i.item) : GetURLItemFromService(i.item);
      return <AccessURLItem key={`accessURL/${i.item.metadata.name}`} URLItem={urlItem} />;
    }
    return;
  }

  private fetchIngresses() {
    // Fetch all related Ingress resources. We don't need to fetch Services as
    // they are expected to be watched by the ServiceTable.
    this.props.ingressRefs.forEach(r => this.props.getResource(r));
  }

  private elemHasItems(i: IKubeItem<IResource | IK8sList<IResource, {}>>) {
    if (i.error) {
      return true;
    }
    if (i.item) {
      const list = i.item as IK8sList<IResource, {}>;
      if (list.items && list.items.length === 0) {
        return false;
      }
      return true;
    }
    return false;
  }

  private hasItems(
    svcs: Array<IKubeItem<IResource | IK8sList<IResource, {}>>>,
    ingresses: Array<IKubeItem<IResource | IK8sList<IResource, {}>>>,
  ) {
    return (
      (svcs.length && svcs.some(svc => this.elemHasItems(svc))) ||
      (ingresses.length && ingresses.some(ingress => this.elemHasItems(ingress)))
    );
  }
}

export default AccessURLTable;
