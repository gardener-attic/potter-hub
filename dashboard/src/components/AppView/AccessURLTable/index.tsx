import AccessURLTable from "./AccessURLTable";

export default AccessURLTable;
