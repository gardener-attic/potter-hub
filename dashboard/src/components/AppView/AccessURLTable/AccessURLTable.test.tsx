import { shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";

import itBehavesLike from "../../../shared/specs";

import ResourceRef from "shared/ResourceRef";
import { IIngressSpec, IResource, IServiceSpec, IServiceStatus } from "../../../shared/types";
import AccessURLItem from "./AccessURLItem";
import AccessURLTable from "./AccessURLTable";

const defaultProps = {
  services: [],
  ingresses: [],
  ingressRefs: [],
  getResource: jest.fn(),
};

describe("when receiving ingresses", () => {
  it("fetches ingresses at mount time", () => {
    const ingress = { name: "ing" } as ResourceRef;
    const mock = jest.fn();
    shallow(<AccessURLTable {...defaultProps} ingressRefs={[ingress]} getResource={mock} />);
    expect(mock).toHaveBeenCalledWith(ingress);
  });

  it("fetches when new ingress refs received", () => {
    const ingress = { name: "ing" } as ResourceRef;
    const mock = jest.fn();
    const wrapper = shallow(<AccessURLTable {...defaultProps} getResource={mock} />);
    wrapper.setProps({ ingressRefs: [ingress] });
    expect(mock).toHaveBeenCalledWith(ingress);
  });
});

context("when fetching ingresses or services", () => {
  itBehavesLike("aLoadingComponent", {
    component: AccessURLTable,
    props: {
      ...defaultProps,
      ingresses: [{ isFetching: true }],
    },
  });
  itBehavesLike("aLoadingComponent", {
    component: AccessURLTable,
    props: {
      ...defaultProps,
      services: [{ isFetching: true }],
    },
  });
});

it("doesn't render anything if the application has no URL", () => {
  const wrapper = shallow(<AccessURLTable {...defaultProps} />);
  expect(wrapper.find("table")).not.toExist();
});

context("when the app contains services", () => {
  it("should omit the Service Table if there are no public services", () => {
    const service = {
      kind: "Service",
      metadata: {
        name: "foo",
      },
      spec: {
        type: "ClusterIP",
        ports: [{ port: 8080 }],
      } as IServiceSpec,
      status: {
        loadBalancer: {},
      } as IServiceStatus,
    } as IResource;
    const services = [{ isFetching: false, item: service }];
    const wrapper = shallow(<AccessURLTable {...defaultProps} services={services} />);
    expect(wrapper.text()).toContain("The current application does not expose a public URL");
  });

  it("should show the table if any service is a LoadBalancer", () => {
    const service = {
      kind: "Service",
      metadata: {
        name: "foo",
      },
      spec: {
        type: "LoadBalancer",
        ports: [{ port: 8080 }],
      } as IServiceSpec,
      status: {
        loadBalancer: {},
      } as IServiceStatus,
    } as IResource;
    const services = [{ isFetching: false, item: service }];
    const wrapper = shallow(<AccessURLTable {...defaultProps} services={services} />);
    expect(wrapper.find(AccessURLItem)).toExist();
    expect(wrapper).toMatchSnapshot();
  });
});

context("when the app contains ingresses", () => {
  it("should show the table with available ingresses", () => {
    const ingress = {
      kind: "Ingress",
      metadata: {
        name: "foo",
      },
      spec: {
        rules: [
          {
            host: "foo.bar",
            http: {
              paths: [{ path: "/ready" }],
            },
          },
        ],
      } as IIngressSpec,
    } as IResource;
    const ingresses = [{ isFetching: false, item: ingress }];
    const wrapper = shallow(<AccessURLTable {...defaultProps} ingresses={ingresses} />);
    expect(wrapper.find(AccessURLItem)).toExist();
    expect(wrapper).toMatchSnapshot();
  });
});

context("when the app contains services and ingresses", () => {
  it("should show the table with available svcs and ingresses", () => {
    const service = {
      kind: "Service",
      metadata: {
        name: "foo",
      },
      spec: {
        type: "LoadBalancer",
        ports: [{ port: 8080 }],
      } as IServiceSpec,
      status: {
        loadBalancer: {},
      } as IServiceStatus,
    } as IResource;
    const services = [{ isFetching: false, item: service }];
    const ingress = {
      kind: "Ingress",
      metadata: {
        name: "foo",
      },
      spec: {
        rules: [
          {
            host: "foo.bar",
            http: {
              paths: [{ path: "/ready" }],
            },
          },
        ],
      } as IIngressSpec,
    } as IResource;
    const ingresses = [{ isFetching: false, item: ingress }];
    const wrapper = shallow(
      <AccessURLTable {...defaultProps} services={services} ingresses={ingresses} />,
    );
    expect(wrapper.find(AccessURLItem)).toExist();
    expect(wrapper).toMatchSnapshot();
  });
});

context("when the app contains resources with errors", () => {
  it("displays the error", () => {
    const services = [{ isFetching: false, error: new Error("could not find Service") }];
    const ingresses = [{ isFetching: false, error: new Error("could not find Ingress") }];
    const wrapper = shallow(
      <AccessURLTable {...defaultProps} services={services} ingresses={ingresses} />,
    );

    // The Service error is not shown, as it is filtered out because without the
    // resource we can't determine whether it is a public LoadBalancer Service
    // or not. The Service error will be shown in the Services table anyway.
    expect(wrapper.find(AccessURLItem)).not.toExist();
    expect(wrapper.find("table").text()).toContain("could not find Ingress");

    expect(wrapper).toMatchSnapshot();
  });
});
