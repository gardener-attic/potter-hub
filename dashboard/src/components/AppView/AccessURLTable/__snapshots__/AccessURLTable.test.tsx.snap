// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when fetching ingresses or services loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
  type={1}
/>
`;

exports[`when fetching ingresses or services loading spinner matches the snapshot 2`] = `
<LoadingWrapper
  loaded={false}
  type={1}
/>
`;

exports[`when the app contains ingresses should show the table with available ingresses 1`] = `
<Fragment>
  <h6>
    Access URLs
  </h6>
  <table>
    <thead>
      <tr>
        <th>
          RESOURCE
        </th>
        <th>
          TYPE
        </th>
        <th>
          URL
        </th>
      </tr>
    </thead>
    <tbody>
      <AccessURLItem
        URLItem={
          Object {
            "URLs": Array [
              "http://foo.bar/ready",
            ],
            "isLink": true,
            "name": "foo",
            "type": "Ingress",
          }
        }
        key="accessURL/foo"
      />
    </tbody>
  </table>
</Fragment>
`;

exports[`when the app contains resources with errors displays the error 1`] = `
<Fragment>
  <h6>
    Access URLs
  </h6>
  <table>
    <thead>
      <tr>
        <th>
          RESOURCE
        </th>
        <th>
          TYPE
        </th>
        <th>
          URL
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="could not find Ingress"
      >
        <td
          colSpan={3}
        >
          Error: 
          could not find Ingress
        </td>
      </tr>
    </tbody>
  </table>
</Fragment>
`;

exports[`when the app contains services and ingresses should show the table with available svcs and ingresses 1`] = `
<Fragment>
  <h6>
    Access URLs
  </h6>
  <table>
    <thead>
      <tr>
        <th>
          RESOURCE
        </th>
        <th>
          TYPE
        </th>
        <th>
          URL
        </th>
      </tr>
    </thead>
    <tbody>
      <AccessURLItem
        URLItem={
          Object {
            "URLs": Array [
              "http://foo.bar/ready",
            ],
            "isLink": true,
            "name": "foo",
            "type": "Ingress",
          }
        }
        key="accessURL/foo"
      />
      <AccessURLItem
        URLItem={
          Object {
            "URLs": Array [
              "Pending",
            ],
            "isLink": false,
            "name": "foo",
            "type": "Service LoadBalancer",
          }
        }
        key="accessURL/foo"
      />
    </tbody>
  </table>
</Fragment>
`;

exports[`when the app contains services should show the table if any service is a LoadBalancer 1`] = `
<Fragment>
  <h6>
    Access URLs
  </h6>
  <table>
    <thead>
      <tr>
        <th>
          RESOURCE
        </th>
        <th>
          TYPE
        </th>
        <th>
          URL
        </th>
      </tr>
    </thead>
    <tbody>
      <AccessURLItem
        URLItem={
          Object {
            "URLs": Array [
              "Pending",
            ],
            "isLink": false,
            "name": "foo",
            "type": "Service LoadBalancer",
          }
        }
        key="accessURL/foo"
      />
    </tbody>
  </table>
</Fragment>
`;
